vendor/bat-native-ledger/src/bat/ledger/internal/database/database_external_transactions.cc

```diff
@@ -0,0 +1,138 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include <utility>
+
+#include "base/strings/string_number_conversions.h"
+#include "base/strings/stringprintf.h"
+#include "bat/ledger/internal/database/database_external_transactions.h"
+#include "bat/ledger/internal/database/database_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+
+namespace ledger::database {
+
+constexpr char kTableName[] = "external_transactions";
+
+DatabaseExternalTransactions::DatabaseExternalTransactions(LedgerImpl* ledger)
+    : DatabaseTable(ledger) {}
+
+DatabaseExternalTransactions::~DatabaseExternalTransactions() = default;
+
+void DatabaseExternalTransactions::Insert(
+    mojom::ExternalTransactionPtr external_transaction,
+    ledger::ResultCallback callback) {
+  if (!external_transaction) {
+    BLOG(0, "external_transaction is null!");
+    return std::move(callback).Run(mojom::Result::LEDGER_ERROR);
+  }
+
+  auto command = mojom::DBCommand::New();
+  command->type = mojom::DBCommand::Type::RUN;
+  command->command = base::StringPrintf(
+      R"(
+        INSERT INTO %s (transaction_id, contribution_id, destination, amount)
+        VALUES(?, ?, ?, ?)
+      )",
+      kTableName);
+  BindString(command.get(), 0, external_transaction->transaction_id);
+  BindString(command.get(), 1, external_transaction->contribution_id);
+  BindString(command.get(), 2, external_transaction->destination);
+  BindString(command.get(), 3, external_transaction->amount);
```

Comment:
is there a possibility of SQL-injection type attacks here? assuming an external party can control part of these values

Reply:
`BindString()` just adds a `DBCommandBindingPtr` to `command`'s `bindings` vector. We [use](https://github.com/brave/brave-core/blob/370015bcdf85b99ace14995a1d5cf68d58f7c1ad/vendor/bat-native-ledger/include/bat/ledger/public/ledger_database.cc#L29) upstream's [`SQLite abstraction layer`](https://source.chromium.org/chromium/chromium/src/+/main:sql/), which in turn uses `sqlite3_bind_*` calls from [`Chromium SQLite`](https://source.chromium.org/chromium/chromium/src/+/main:third_party/sqlite/) (where the actual bindings happen).